apis/duck/v1beta1/source_types.go

```diff
@@ -0,0 +1,174 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"context"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/apis/duck"
+	apisv1alpha1 "knative.dev/pkg/apis/v1alpha1"
+)
+
+// Source is an Implementable "duck type".
+var _ duck.Implementable = (*Source)(nil)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// Source is the minimum resource shape to adhere to the Source Specification.
+// This duck type is intended to allow implementors of Sources and
+// Importers to verify their own resources meet the expectations.
+// This is not a real resource.
+// NOTE: The Source Specification is in progress and the shape and names could
+// be modified until it has been accepted.
+type Source struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   SourceSpec   `json:"spec"`
+	Status SourceStatus `json:"status"`
+}
+
+type SourceSpec struct {
+	// Sink is a reference to an object that will resolve to a domain name or a
+	// URI directly to use as the sink.
+	Sink apisv1alpha1.Destination `json:"sink,omitempty"`
+
+	// CloudEventOverrides defines overrides to control the output format and
+	// modifications of the event sent to the sink.
+	// +optional
+	CloudEventOverrides CloudEventOverrides `json:"ceOverrides,omitempty"`
+}
+
+type CloudEventOverrides struct {
+	// OutputFormat describes the CloudEvent output format the source
+	// should send events in. All formats are over HTTP.
+	// Defaults to binary.
+	// +optional
+	OutputFormat OutputFormatType `json:"outputFormat,omitempty"`
```

Comment:
I worry that the presence of this field, even if it's optional and a SHOULD, will practically force importers to implement both HTTP output formats. Do we actually need this in the spec? Can it be an annotation until we know that this is a common need?

Reply:
producing binary or structured is the min bar for an SDK. I think it is reasonable to assume the source impl can do this.

Reply:
That does seem reasonable! But I still don't understand why we need this in the spec.

The [HTTP transport binding spec](https://github.com/cloudevents/spec/blob/master/http-transport-binding.md#3-http-message-mapping) implies that all receivers must be able to handle both binary and structured modes. If we expect receivers to be CloudEvents compliant (I think we do), then in what scenario would the user ever need to set this field?